File: expandabletext/src/main/java/com/kolyall/view/expandabletext/ExpandableTextView.java
package com.kolyall.view.expandabletext;

import android.content.Context;
import android.content.res.TypedArray;
import android.support.v7.widget.AppCompatTextView;
import android.text.SpannableStringBuilder;
import android.util.AttributeSet;
import android.view.View;

import lombok.Getter;
import lombok.Setter;
import lombok.experimental.Accessors;

/**
 * Created by author on 13.01.2018.
 */
@Accessors(prefix = "m")
public class ExpandableTextView extends AppCompatTextView implements View.OnClickListener {
    private static final int DEFAULT_TRIM_LENGTH = 160;
    private static final String ELLIPSIS = "...";

    private CharSequence originalText;
    private CharSequence trimmedText;
    private BufferType bufferType;
    @Getter private boolean mTrim = true;
    private int trimLength;

    public ExpandableTextView(Context context) {
        this(context, null);
    }

    public ExpandableTextView(Context context, AttributeSet attrs) {
        super(context, attrs);

        TypedArray typedArray = context.obtainStyledAttributes(attrs, R.styleable.ExpandableTextView);
        this.trimLength = typedArray.getInt(R.styleable.ExpandableTextView_trimLength, DEFAULT_TRIM_LENGTH);
        typedArray.recycle();

        setOnClickListener(this);
    }

    private void setText() {
        super.setText(getDisplayableText(), bufferType);
    }

    private CharSequence getDisplayableText() {
        return mTrim ? trimmedText : originalText;
    }

    @Override
    public void setText(CharSequence text, BufferType type) {
        originalText = text;
        trimmedText = getTrimmedText(text);
        bufferType = type;
        setText();
    }

    private CharSequence getTrimmedText(CharSequence text) {
        if (isModified()) {
            return new SpannableStringBuilder(originalText, 0, trimLength + 1).append(ELLIPSIS);
        } else {
            return originalText;
        }
    }

    public boolean isModified() {
        return originalText != null && originalText.length() > trimLength;
    }

    public CharSequence getOriginalText() {
        return originalText;
    }

    public void setTrimLength(int trimLength) {
        this.trimLength = trimLength;
        trimmedText = getTrimmedText(originalText);
        setText();
    }

    public int getTrimLength() {
        return trimLength;
    }

    @Override
    public void onClick(View v) {
        mTrim = !mTrim;
        setText();
        mOnExpandableTextViewListener.onExpandChanged(!mTrim);
    }

    @Setter OnExpandableTextViewListener mOnExpandableTextViewListener;

    public interface OnExpandableTextViewListener{
        void onExpandChanged(boolean expanded);
    }
}

File: expandabletext/src/main/java/com/kolyall/view/expandabletext/ExpandableTextView.kt
package com.kolyall.view.expandabletext

import android.content.Context
import android.text.SpannableStringBuilder
import android.util.AttributeSet
import android.view.View
import androidx.appcompat.widget.AppCompatTextView

class ExpandableTextView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null
) : AppCompatTextView(context, attrs), View.OnClickListener {
    private var originalText: CharSequence? = null
    private var trimmedText: CharSequence? = null
    private var bufferType: BufferType? = null
    var isTrim = true
    private var trimLength: Int

    private fun setText() {
        super.setText(displayableText, bufferType)
    }

    private val displayableText: CharSequence?
        get() = if (isTrim) trimmedText else originalText

    override fun setText(text: CharSequence, type: BufferType) {
        originalText = text
        trimmedText = getTrimmedText(text)
        bufferType = type
        setText()
    }

    private fun getTrimmedText(text: CharSequence?): CharSequence? {
        return if (isModified) {
            SpannableStringBuilder(
                originalText,
                0,
                trimLength + 1
            ).append(ELLIPSIS)
        } else {
            originalText
        }
    }

    val isModified: Boolean
        get() = originalText != null && originalText!!.length > trimLength

    fun setTrimLength(trimLength: Int) {
        this.trimLength = trimLength
        trimmedText = getTrimmedText(originalText)
        setText()
    }

    fun getTrimLength(): Int {
        return trimLength
    }

    override fun onClick(v: View) {
        isTrim = !isTrim
        setText()
        onExpandableTextViewListener?.onExpandChanged(!isTrim)
    }

    var onExpandableTextViewListener: OnExpandableTextViewListener? = null

    interface OnExpandableTextViewListener {
        fun onExpandChanged(expanded: Boolean)
    }

    companion object {
        private const val DEFAULT_TRIM_LENGTH = 160
        private const val ELLIPSIS = "..."
    }

    init {
        val typedArray =
            context.obtainStyledAttributes(attrs, R.styleable.ExpandableTextView)
        trimLength = typedArray.getInt(
            R.styleable.ExpandableTextView_trimLength,
            DEFAULT_TRIM_LENGTH
        )
        typedArray.recycle()
        setOnClickListener(this)
    }
}